src/callbacks.cpp

```diff
@@ -0,0 +1,231 @@
+
+            RCLCPP_INFO(this->get_logger(),
+                        "WaypointNode::%s: --- NODE DEACTIVATED ---", __func__);
+        }
+    }
+
+    // If the node should be deactivated, don't parse the payload as it is only
+    // relevant in active mode
+    if (!this->get_active()) {
+        return;
+    }
+
+    // Parse payload
+    nlohmann::json cmd_json;
+    try {
+        cmd_json = common_lib::CommandDefinitions::parse_check_json_str(
+            msg.payload,
+            common_lib::CommandDefinitions::get_waypoint_command_definition());
+    } catch (const std::runtime_error &e) {
+        RCLCPP_FATAL(this->get_logger(),
+                     "WaypointNode::%s: Received invalid json as "
+                     "payload: %s",
+                     __func__, e.what());
+        this->job_finished(
+            (std::string) "WaypointNode::" + (std::string) __func__ +
+            ": Received "
+            "invalid json as payload: " +
+            e.what());
+        reset_node();
+        return;
+    }
+
+    // Set flag to true to indicate that the object has values in it
+    cmd.values_set = true;
+
+    // Required parameters
+    cmd.target_coordinate_lat = cmd_json.at("target_coordinate_lat");
+    cmd.target_coordinate_lon = cmd_json.at("target_coordinate_lon");
+    cmd.cruise_height_cm = cmd_json.at("cruise_height_cm");
+    cmd.target_height_cm = cmd_json.at("target_height_cm");
+    cmd.horizontal_speed_mps = cmd_json.at("horizontal_speed_mps");
+    cmd.vertical_speed_mps = cmd_json.at("vertical_speed_mps");
+
+    // Optional parameters
+    if (cmd_json.contains("pre_wait_time_ms"))
+        cmd.pre_wait_time_ms = cmd_json.at("pre_wait_time_ms");
+
+    if (cmd_json.contains("post_wait_time_ms"))
+        cmd.post_wait_time_ms = cmd_json.at("post_wait_time_ms");
+}
+
+/**
+ * @brief Callback function for receiving GPS position messages.
+ *
+ * This function is called when a GPS position message is received. It extracts
+ * the latitude, longitude, and height information from the message and updates
+ * the corresponding variables in the `pos` object.
+ *
+ * @param msg The GPS position message containing the latitude, longitude, and
+ * height information.
+ */
+void WaypointNode::callback_position(const interfaces::msg::GPSPosition &msg) {
+    // Store current timestamp for later
+    rclcpp::Time timestamp_now = this->now();
+
+    RCLCPP_DEBUG(this->get_logger(),
+                 "WaypointNode::%s: Received position from "
+                 "'%s': lat: %f, lon: %f, height: %f",
+                 __func__, msg.sender_id.c_str(), msg.latitude_deg,
+                 msg.longitude_deg, msg.relative_altitude_m);
+
+    // Check timestamp
+    if (timestamp_now - rclcpp::Time(msg.time_stamp) >
+        rclcpp::Duration(std::chrono::duration<int64_t, std::milli>(
+            max_position_msg_time_difference_ms))) {
+        // Message is too old, react based on the current 'active' state
```

Comment:
Maybe check the sender id as well.

Reply:
I don't see that as a problem.
If someone else is sending on the topic we have an intruder (by accident or on purpose) and therefore way bigger problems anyway.

Reply:
I think its important to detects these things, the problem should not become even bigger because we let it.

Reply:
We cannot check everything. Time is limited so we need to focus on the things that matter.

Reply:
Your node your call